p2p/src/peer_manager/peerdb/address_tables/table.rs

```diff
@@ -0,0 +1,532 @@
+// Copyright (c) 2021-2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{
+    collections::{btree_map, BTreeMap},
+    hash::{Hash, Hasher},
+};
+
+use p2p_types::socket_address::SocketAddress;
+
+use crate::peer_manager::address_groups::AddressGroup;
+
+use super::RandomKey;
+
+macro_rules! calc_hash {
+    ($($val:expr),+) => {
+        {
+            let mut hasher = std::collections::hash_map::DefaultHasher::new();
+            $(
+                $val.hash(&mut hasher);
+            )*
+            hasher.finish()
+        }
+    };
+}
```

Comment:
I didn't realize I could use tuples to simulate a variadic function :-).
So, I replaced the macro with a function:
```
/// Calculate hash of a 'single' value.
/// 
/// Note: if multiple values are required to produce a single hash, pass them as a tuple.
pub fn calc_hash<T: Hash>(val: &T) -> u64 {
    let mut hasher = std::collections::hash_map::DefaultHasher::new();
    val.hash(&mut hasher);
    hasher.finish()
}
```
Using `BuildHasher::hash_one` IMO is redundant, because it's trivial and just introduces another level of indirection.